### PR TITLE
Closes #1228 produce bmi records for more visits

### LINKED ISSUE #1228: Feature Request:  Produce BMI records for more visits with derive_param_bmi()
### Feature Idea

I expect `derive_param_bmi` to generate BMI records for each weight records of USUBJID, VISITNUM, ADT, and VSTPT in ADVS. However, it only generate the BMI records correponding to the height records. Taking the admiral_advs dataset in the admiral package (0.7.0) as an example, I expect to see 4309 BMI records, just like WEIGHT, not 508 records as HEIGHT. Thanks

### Relevant Input

_No response_

### Relevant Output

_No response_

### Reproducible Example/Pseudo Code

_No response_

### PR DESCRIPTION
**Pull request by jeffreyad**
…parameter

Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiraldev/devel/articles/development_process.html) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/devel/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/devel/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/cran-release/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by jeffreyad** [comment]
Added `constant_height` parameter to `derive_param_bmi()` and `derive_param_bsa()`.  Default value is `FALSE` so function will remain consistent unless changed to `TRUE`.  If `TRUE` then `constant_by_vars` is expected.

```
advs <- tribble(
  ~USUBJID,      ~PARAMCD, ~PARAM,        ~AVAL, ~AVISIT,
  "01-701-1015", "HEIGHT", "Height (cm)", 147,   "SCREENING",
  "01-701-1015", "WEIGHT", "Weight (kg)", 54.0,  "SCREENING",
  "01-701-1015", "WEIGHT", "Weight (kg)", 54.4,  "BASELINE",
  "01-701-1015", "WEIGHT", "Weight (kg)", 53.1,  "WEEK 2",
  "01-701-1028", "HEIGHT", "Height (cm)", 163,   "SCREENING",
  "01-701-1028", "WEIGHT", "Weight (kg)", 78.5,  "SCREENING",
  "01-701-1028", "WEIGHT", "Weight (kg)", 80.3,  "BASELINE",
  "01-701-1028", "WEIGHT", "Weight (kg)", 80.7,  "WEEK 2"
)

# Derive BMI where height is measured only once
derive_param_bmi(
  advs,
  by_vars = exprs(USUBJID, AVISIT),
  weight_code = "WEIGHT",
  height_code = "HEIGHT",
  set_values_to = exprs(
    PARAMCD = "BMI",
    PARAM = "Body Mass Index (kg/m^2)"
  ),
  get_unit_expr = extract_unit(PARAM),
  constant_height = TRUE,
  constant_by_vars = exprs(USUBJID)
)
#> # A tibble: 14 × 5
#>    USUBJID     PARAMCD PARAM                     AVAL AVISIT   
#>    <chr>       <chr>   <chr>                    <dbl> <chr>    
#>  1 01-701-1015 HEIGHT  Height (cm)              147   SCREENING
#>  2 01-701-1015 WEIGHT  Weight (kg)               54   SCREENING
#>  3 01-701-1015 WEIGHT  Weight (kg)               54.4 BASELINE 
#>  4 01-701-1015 WEIGHT  Weight (kg)               53.1 WEEK 2   
#>  5 01-701-1028 HEIGHT  Height (cm)              163   SCREENING
#>  6 01-701-1028 WEIGHT  Weight (kg)               78.5 SCREENING
#>  7 01-701-1028 WEIGHT  Weight (kg)               80.3 BASELINE 
#>  8 01-701-1028 WEIGHT  Weight (kg)               80.7 WEEK 2   
#>  9 01-701-1015 BMI     Body Mass Index (kg/m^2)  25.0 SCREENING
#> 10 01-701-1015 BMI     Body Mass Index (kg/m^2)  25.2 BASELINE 
#> 11 01-701-1015 BMI     Body Mass Index (kg/m^2)  24.6 WEEK 2   
#> 12 01-701-1028 BMI     Body Mass Index (kg/m^2)  29.5 SCREENING
#> 13 01-701-1028 BMI     Body Mass Index (kg/m^2)  30.2 BASELINE 
#> 14 01-701-1028 BMI     Body Mass Index (kg/m^2)  30.4 WEEK 2 
```

File: R/derive_advs_params.R [hunk 1]
```diff
@@ -255,6 +255,15 @@ compute_map <- function(diabp, sysbp, hr = NULL) {
...
 #'   Permitted Values: character value
 #'
+#' @param constant_height HEIGHT parameter is constant
+#'
+#'   If the HEIGHT value is considered constant (e.g. measured once at screening),
+#'   the values of BSA will be calculated for each WEIGHT value.
+#'
+#'   Permitted Values: logical scalar
+#'
+#'   Default Values: FALSE
```
Comment (manciniedoardo) [inline R/derive_advs_params.R:265]: I think the convention now in {admiral} is to not specify the default value in the roxygen header.

File: R/derive_advs_params.R [hunk 2]
```diff
@@ -255,6 +255,15 @@ compute_map <- function(diabp, sysbp, hr = NULL) {
...
+#'
```
Comment (manciniedoardo) [inline R/derive_advs_params.R:266]: Same comments as for other function.

File: R/derive_advs_params.R [hunk 3]
```diff
@@ -255,6 +255,15 @@ compute_map <- function(diabp, sysbp, hr = NULL) {
 #'
 #'   Permitted Values: character value
 #'
+#' @param constant_height HEIGHT parameter is constant
+#'
+#'   If the HEIGHT value is considered constant (e.g. measured once at screening),
+#'   the values of BSA will be calculated for each WEIGHT value.
```
Comment (manciniedoardo) [inline R/derive_advs_params.R:261]: I would rephrase to: "To calculate BSA at every visit at which WEIGHT is collected using the patient's constant HEIGHT, set to TRUE. Otherwise, to calculate BSA only for the first instance in which HEIGHT and WEIGHT are collected, set to FALSE.

File: R/derive_advs_params.R [hunk 4]
```diff
@@ -255,6 +255,15 @@ compute_map <- function(diabp, sysbp, hr = NULL) {
 #'
 #'   Permitted Values: character value
 #'
+#' @param constant_height HEIGHT parameter is constant
```
[the remark this answers is not in the document]
Comment (bundfussr) [reply R/derive_advs_params.R:258]: I think we always want to derive BMI/BSA for all visits.

However, I wonder if we need this argument at all. Providing `constant_by_vars` should be sufficient. If it is not `NULL`, `constant_parameters` and `constant_by_vars` is specified in the `derive_param_computed()` call. Otherwise, these are not specified.

File: R/derive_advs_params.R [hunk 5]
```diff
@@ -530,13 +584,30 @@ compute_bsa <- function(height = height,
 #'   ),
 #'   get_unit_expr = extract_unit(PARAM)
 #' )
+#'
```
Comment (bundfussr) [inline R/derive_advs_params.R:587]: Could you update the first example (for BMI and BSA) such that height is measured at every visit? The current example does not make much sense.

I would also change the order of the examples because measuring height only once is the more common case.

**Review by bundfussr — CHANGES_REQUESTED** [review]
Please add an item to `NEWS.md`.

File: R/derive_advs_params.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeffreyad) [reply R/derive_advs_params.R]: @bundfussr I like the idea of using `constant_by_vars`.  This is updated now.  Thanks!

File: R/derive_advs_params.R (hunk 5 above)
Comment (jeffreyad) [reply R/derive_advs_params.R:612]: @bundfussr updated order of examples, thanks!

File: R/derive_advs_params.R [hunk 6]
```diff
@@ -530,13 +584,30 @@ compute_bsa <- function(height = height,
...
+#' # Example 2: Derive BMI where height is measured only once
```
[the remark this answers is not in the document]
Comment (jeffreyad) [reply R/derive_advs_params.R]: @manciniedoardo updated examples to be clearer, thanks!

File: R/derive_advs_params.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (manciniedoardo) [reply R/derive_advs_params.R]: I like this idea also 😄

**Review by manciniedoardo — APPROVED** [review]
Thanks for working on this @jeffreyad !

**Review by bms63 — APPROVED** [review]
I think we should document `constant_by_vars` in the programming strategy.  We might make use of this idea again and nice to be consistent.  @bundfussr what do you think?

**Comment by bundfussr** [comment]
> I think we should document `constant_by_vars` in the programming strategy. We might make use of this idea again and nice to be consistent. @bundfussr what do you think?

@bms63 , I am not sure. Maybe we should wait for the results of the clean-up. One option to discuss is that we just keep `derive_param_computed()` and the `compute_*()` functions. Then `constant_by_vars` would be used by one function only and we would not need to add it to the programming strategy.

File: R/derive_advs_params.R [hunk 7]
```diff
@@ -239,21 +239,33 @@ compute_map <- function(diabp, sysbp, hr = NULL) {
...
-#'   Permitted Values: character value
+#'   *Permitted Values:* character value
+#'
+#' @param constant_by_vars By variables for when HEIGHT is constant
+#'
+#'   When HEIGHT is constant, the HEIGHT parameters (measured only once) are merged
+#'   to the other parameters using the specified variables.
+#'
+#'   If height is constant (e.g. only measured once at screening or baseline) then
+#'   use `constant_by_vars` to select the subject-level variable to merge on (e.g. USUBJID).
```
Comment (bundfussr) [inline R/derive_advs_params.R:264]: Please use backticks for `USUBJID`.

File: R/derive_advs_params.R [hunk 8]
```diff
@@ -477,14 +541,28 @@ compute_bsa <- function(height = height,
...
+#'
+#'   *Permitted Values:* logical scalar
+#'
+#' @param constant_by_vars By variables for when HEIGHT is constant
+#'
+#'   When HEIGHT is constant, the HEIGHT parameters (measured only once) are merged
+#'   to the other parameters using the specified variables.
+#'
+#'   If height is constant (e.g. only measured once at screening or baseline) then
+#'   use `constant_by_vars` to select the subject-level variable to merge on (e.g. USUBJID).
```
Comment (bundfussr) [inline R/derive_advs_params.R:561]: Please use backtitcks for `USUBJID`.

**Review by bundfussr — CHANGES_REQUESTED** [review]
Sorry, for the late feedback. I was busy with other projects.

File: R/derive_advs_params.R (hunk 8 above)
Comment (jeffreyad) [reply R/derive_advs_params.R]: Added

**Review by bundfussr — APPROVED** [review]

File: R/derive_advs_params.R (hunk 7 above)
Comment (jeffreyad) [reply R/derive_advs_params.R]: Added
